### PR TITLE
Implement is_observable rules in scenario

### PR DESCRIPTION
**Pull request by mrkickling**
Gives the option to add observability rules into scenario configs.

- User can now specify observability rules for certain attack steps either by
  - asset name + attack step name
  - asset type + attack step name
 
Scenario loading:
- The observability rules are applied when loading a scenario, this updates each node.extras['observable'] to either 1 or 0
  - 1 if the node follows one of the rules given, otherwise 0 
- If no rules are given in the scenario config file, all steps/nodes are marked as observable  = 1.

Observability rules are validated to make sure they contain valid asset types/names and and attack steps.

MalSimulator:
- The 'is_observable' in the observation given from the simulator will contain 1 or 0 depending on if rule was created in scenario config file. (previously it was always 1)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -171,4 +171,4 @@ cython_debug/
 #.idea/
 
 # Allow .json/.yml files in testdata
-!tests/testdata/*
+!tests/testdata/*/*
```
[the remark this answers is not in the document]
Comment (mrkickling) [reply .gitignore:174]: The yaml files that were inside of subfolders

**Review by andrewbwm — APPROVED** [review]

File: .gitignore (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrkickling) [reply .gitignore]: Changed into /**

File: README.md [hunk 2]
```diff
@@ -24,7 +24,8 @@ attacker_agent_class: 'BreadthFirstAttacker' | 'DepthFirstAttacker' | 'KeyboardA
 # For defender_agent_class, null and False are treated the same - no defender will be used in the simulation
 defender_agent_class: 'BreadthFirstAttacker' | 'DepthFirstAttacker' | 'KeyboardAgent' | null | False
 
-# Rewards for each attack step
+
+# Optionally add rewards for each attack step
 rewards:
   <full name of attack step>: <reward>
   <full name of attack step>: <reward>
```
[the remark this answers is not in the document]
Comment (mrkickling) [reply README.md]: Changed it to be just one line, and added two lines to example instead.

File: README.md [hunk 3]
```diff
@@ -43,6 +45,37 @@ attacker_entry_points:
...
+# If 'observable_attack_steps' are not set:
+# - All nodes will be marked as observable
+#
+observable_attack_steps:
+  by_asset_type:
+    <ASSET TYPE>:
+      - <ATTACK STEP NAME>
+  by_asset_name:
+    <ASSET NAME>:
+      - <ATTACK STEP NAME>
```
[the remark this answers is not in the document]
Comment (mrkickling) [reply README.md]: Fixed

File: README.md [hunk 4]
```diff
@@ -43,6 +45,37 @@ attacker_entry_points:
...
+
+  # Example:
+  # observable_attack_steps:
+  #   by_asset_type:
+  #     Host:
+  #       - access
+  #     Data:
+  #       - read
```
[the remark this answers is not in the document]
Comment (mrkickling) [reply README.md]: Done

File: tests/testdata/scenarios/simple_filtered_observability_scenario.yml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+# If observable_attack_steps are set, all attack steps not
+# matching these rules will have observability set to False.
+observable_attack_steps:
+  by_asset_type:
+    Application:
+      - fullAccess
+
+  by_asset_name:
+    OS App:
+      - fullAccess
```
[the remark this answers is not in the document]
Comment (mrkickling) [reply tests/testdata/scenarios/simple_filtered_observability_scenario.yml]: changed the second one to User assume instead

File: tests/testdata/scenarios/traininglang_observability_scenario.yml [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+observable_attack_steps:
+  by_asset_type:
+    Host:
+      - access
+    Data:
+      - read
+
+  by_asset_name:
+    User:3:
+      - phishing
```
[the remark this answers is not in the document]
Comment (mrkickling) [reply tests/testdata/scenarios/traininglang_observability_scenario.yml]: fixed

**Pull request merged**